Fill.py

```diff
@@ -12,30 +14,35 @@ class FillError(RuntimeError):
     pass
 
 
-def fill_restrictive(world: MultiWorld, base_state: CollectionState, locations, itempool, single_player_placement=False,
-                     lock=False):
-    def sweep_from_pool():
-        new_state = base_state.copy()
-        for item in itempool:
-            new_state.collect(item, True)
-        new_state.sweep_for_events()
-        return new_state
+def sweep_from_pool(base_state: CollectionState, itempool):
+    new_state = base_state.copy()
+    for item in itempool:
+        new_state.collect(item, True)
+    new_state.sweep_for_events()
+    return new_state
+
 
+def fill_restrictive(world: MultiWorld, base_state: CollectionState, locations, itempool: list[Item], single_player_placement=False,
+                     lock=False):
     unplaced_items = []
     placements = []
 
-    reachable_items = {}
+    swapped_items = Counter()
+    reachable_items: dict[str, deque] = {}
     for item in itempool:
-        reachable_items.setdefault(item.player, []).append(item)
+        reachable_items.setdefault(item.player, deque()).append(item)
 
     while any(reachable_items.values()) and locations:
-        items_to_place = [items.pop() for items in reachable_items.values() if items]  # grab one item per player
+        # grab one item per player
+        items_to_place = [items.pop()
+                          for items in reachable_items.values() if items]
         for item in items_to_place:
             itempool.remove(item)
-        maximum_exploration_state = sweep_from_pool()
+        maximum_exploration_state = sweep_from_pool(base_state, itempool)
         has_beaten_game = world.has_beaten_game(maximum_exploration_state)
 
         for item_to_place in items_to_place:
+            spot_to_fill: Location = None
             if world.accessibility[item_to_place.player] == 'minimal':
                 perform_access_check = not world.has_beaten_game(maximum_exploration_state,
```

Comment:
I believe this condition prevents us from checking the rules for this location, which allows us to place an item in an unreachable location if the game is still beatable. Since this happens before we try to swap item locations, the new swapping code shouldn't prevent us from placing unreachable items